### ISSUE TITLE
Feature Request: Pass CStruct by value

### ISSUE THREAD

**Issue by locriacyber** [rakudo/rakudo#4970]
## The Problem
Currently, there is no way to pass struct by value to C function with `NativeCall`.

For example:
```
#include <stdio.h>

typedef struct {
    int a;
    int b[3];
    int c;
} InlinedArrayInStruct;

void print_struct_by_pointer(InlinedArrayInStruct *ss) {
    InlinedArrayInStruct s = *ss;
    printf("%d, (%d, %d, %d), %d\n", s.a, s.b[0], s.b[1], s.b[2], s.c);
}

void print_struct_by_value(InlinedArrayInStruct s) {
    printf("%d, (%d, %d, %d), %d\n", s.a, s.b[0], s.b[1], s.b[2], s.c);
}
```

Currently, there is no way to call `print_struct_by_value` in the language.

Related issues with CStruct: #4969 #4961 https://github.com/rakudo/rakudo/issues/4022

## Environment
* Compiler version (`perl6 -v` or `raku -v`): 2022-06

**Comment by locriacyber** [comment]
@niner people over IRC told me to ask you about NativeCall.

From what I understand, I think we need a new (not backwards-compatible) FFI for Raku that allows passing by reference. It is more orthogonal to remove `HAS` and make `has` mean "include by value".

That aside, I want to help writing the new FFI for Raku (using libffi). Please give me guidance on how the current NativeCall works in Raku. I tracked it down to codegen in MoarVM, and then I was lost. How does MoarVM use libc anyway?

**Comment by patrickbkr** [comment]
@locriacyber Did you ever hear back? (over IRC or some other means)

**Comment by locriacyber** [comment]
No

**Comment by patrickbkr** [comment]
I dug out a comment about this here: https://irclogs.raku.org/raku-dev/2021-01-23.html#12:53

@niner: Can you give a "A NativeCall:ver<2> is the right approach and something worth trying." vs "It's complicated." assessment?

**Comment by niner** [comment]
Yes, I think this is `ver<2>` material. Thing is, there is a lot wrong in NativeCall (despite it working so well in so many cases). If we're going to version 2, I'd like to fix them at the same time. E.g. memory management is broken quite fundamentally leading to segfaults when Rakudo tries to free a nested data structure. The whole `managed` business is also cumbersome to use and very inefficient. String encodings are quite broken (luckily UTF-8 is so ubiquitous that you don't notice this often). It would be lovely to be able to reuse a library handle or just be able to check if a library could be loaded - easy to implement but we'd have to agree on an API design first. A lot of work has already been done on these in several branches. The memory management cleanup branch could benefit nicely from RakuAST which is where I spend 100 % of my Raku time currently.

**Comment by patrickbkr** [comment]
Thanks for the insights! So I guess this is in the it's-complicated category.

**Comment by ab5tract** [comment]
This one is definitely a missing puzzle piece WRT otherwise very easy C ABI integrations.
